### PR TITLE
Fix #7048: PCP heal path dependent types

### LINKED ISSUE #7048: Path dependent types don't play well with macros and typeclasses
Consider the following:

```scala
import scala.quoted._


trait IsExpr[T] {
  type Underlying
  def toExpr(x: T): Expr[Underlying]
}

// object IsExpr { type Aux[T, U] = IsExpr[T] { type Underlying = U } }

// given [U] as IsExpr.Aux[Expr[U], U] = new IsExpr[Expr[U]] {
given [U] as IsExpr[Expr[U]] = new IsExpr[Expr[U]] {
  type Underlying = U
  def toExpr(x: Expr[U]): Expr[U] = x
}

def f(x: Any): String = x.toString

def g[T](x: T) given (e: IsExpr[T], tu: Type[e.Underlying]): given QuoteContext => Expr[String] = {
  val underlying: Expr[e.Underlying] = e.toExpr(x)
  '{f($underlying)}
}

// def g[T, U](x: T) given (e: IsExpr.Aux[T, U], tu: Type[U]): given QuoteContext => Expr[String] = {
//   val underlying: Expr[U] = e.toExpr(x)
//   '{f($underlying)}
// }

inline def mcr: Any = ${mcrImpl}
def mcrImpl given QuoteContext: Expr[Any] = {
  val x = '{1}
  g(x)
}
```

Will output:

```scala
-- Error: ../pg/Lib.scala:21:6 -------------------------------------------------
21 |  '{f($underlying)}
   |      ^^^^^^^^^^^
   |      access to value e from wrong staging level:
   |       - the definition is at level 0,
   |       - but the access is at level 1.
```

Sort of makes sense, since `e` is indeed defined at level 0. However, two facts make me think something can be done here. First, only the path-dependent type is accessed there (presumably, since the compiler needs to know the type of `underlying`), and we have it with the `Type` argument. Second, the following trick works:

```scala
import scala.quoted._


trait IsExpr[T] {
  type Underlying
  def toExpr(x: T): Expr[Underlying]
}

object IsExpr { type Aux[T, U] = IsExpr[T] { type Underlying = U } }

given [U] as IsExpr.Aux[Expr[U], U] = new IsExpr[Expr[U]] {
// given [U] as IsExpr[Expr[U]] = new IsExpr[Expr[U]] {
  type Underlying = U
  def toExpr(x: Expr[U]): Expr[U] = x
}

def f(x: Any): String = x.toString

// def g[T](x: T) given (e: IsExpr[T], tu: Type[e.Underlying]): given QuoteContext => Expr[String] = {
//   val underlying: Expr[e.Underlying] = e.toExpr(x)
//   '{f($underlying)}
// }

def g[T, U](x: T) given (e: IsExpr.Aux[T, U], tu: Type[U]): given QuoteContext => Expr[String] = {
  val underlying: Expr[U] = e.toExpr(x)
  '{f($underlying)}
}

inline def mcr: Any = ${mcrImpl}
def mcrImpl given QuoteContext: Expr[Any] = {
  val x = '{1}
  g(x)
}
```

`Aux` trick comes from Shapeless 2 and was used to bypass the limitations of Scala 2, where it was impossible to refer to a path-dependent type of a sibling argument from the same argument list.

### PR DESCRIPTION
**Pull request by nicolasstucki**
* Allow path dependent type if their prefix is statically known. As in `'[p.X]` where `p` is statically known and therefore is phase consistent.
* Type heal path dependent type with paths that are not phase consistent. As in `'{ val x: p.X = ... }` where `p` is not phase consistent but there is a given `Type[p.X]` in scope.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala [hunk 1]
```diff
@@ -144,7 +144,17 @@ class PCPCheckAndHeal(@constructorOnly ictx: Context) extends TreeMapWithStages(
...
     def isClassRef = sym.isClass && !tp.isInstanceOf[ThisType]
 
-    if (!sym.exists || levelOK(sym))
+    /** Is this a static path or a type porjection with a static prefix */
+    def isStaticPathOK(tp1: Type): Boolean = tp1 match {
+      case tp1: TypeRef => tp1.symbol.is(Package) || isStaticPathOK(tp1.prefix)
+      case tp1: TermRef => tp1.symbol.isStaticOwner || isStaticPathOK(tp1.prefix)
+      case tp1: ThisType => tp1.tref.symbol.is(Package) || tp1.tref.symbol.is(Module)
+      case tp1: AppliedType => isStaticPathOK(tp1.tycon)
+      case _: SkolemType => true
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: This skolem type appeared in the dependent parameter in https://github.com/lampepfl/dotty/blob/master/tests/pending/run-macros/i7048/Lib_1.scala#L23

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala [hunk 2]
```diff
@@ -144,7 +144,17 @@ class PCPCheckAndHeal(@constructorOnly ictx: Context) extends TreeMapWithStages(
     /** Is a reference to a class but not `this.type` */
     def isClassRef = sym.isClass && !tp.isInstanceOf[ThisType]
 
-    if (!sym.exists || levelOK(sym))
+    /** Is this a static path or a type porjection with a static prefix */
+    def isStaticPathOK(tp1: Type): Boolean = tp1 match {
+      case tp1: TypeRef => tp1.symbol.is(Package) || isStaticPathOK(tp1.prefix)
+      case tp1: TermRef => tp1.symbol.isStaticOwner || isStaticPathOK(tp1.prefix)
+      case tp1: ThisType => tp1.tref.symbol.is(Package) || tp1.tref.symbol.is(Module)
+      case tp1: AppliedType => isStaticPathOK(tp1.tycon)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: Applied types appear with type projections such as in https://github.com/lampepfl/dotty/blob/master/tests/pos/i5954c.scala#L9

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala [hunk 3]
```diff
@@ -144,7 +144,17 @@ class PCPCheckAndHeal(@constructorOnly ictx: Context) extends TreeMapWithStages(
     /** Is a reference to a class but not `this.type` */
     def isClassRef = sym.isClass && !tp.isInstanceOf[ThisType]
 
-    if (!sym.exists || levelOK(sym))
+    /** Is this a static path or a type porjection with a static prefix */
+    def isStaticPathOK(tp1: Type): Boolean = tp1 match {
+      case tp1: TypeRef => tp1.symbol.is(Package) || isStaticPathOK(tp1.prefix)
+      case tp1: TermRef => tp1.symbol.isStaticOwner || isStaticPathOK(tp1.prefix)
+      case tp1: ThisType => tp1.tref.symbol.is(Package) || tp1.tref.symbol.is(Module)
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: Typo: "porjection"

I don't think this is the right logic for isStaticPathOK. How about:
```
tp1.stripTypeVar match
  case tp1: TypeRef => tp1.symbol.is(Package) || isStaticPathOK(tp1.prefix)
  case tp1: TermRef => tp1.symbol.is(Module) && tp1.symbol.isStatic
  case tp1: ThisType => tp1.cls.isStaticOwner
  case tp1: AppliedType => isStaticPathOK(tp1.tycon)

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala [hunk 4]
```diff
@@ -171,7 +181,7 @@ class PCPCheckAndHeal(@constructorOnly ictx: Context) extends TreeMapWithStages(
             sym.isClass // reference to this in inline methods
           )
     case None =>
-      !sym.is(Param) || levelOK(sym.owner)
+      sym.is(Package)|| sym.owner.isStaticOwner || levelOK(sym.owner)
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: Missing space

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala [hunk 5]
```diff
@@ -144,7 +144,17 @@ class PCPCheckAndHeal(@constructorOnly ictx: Context) extends TreeMapWithStages(
...
+    /** Is this a static path or a type porjection with a static prefix */
+    def isStaticPathOK(tp1: Type): Boolean = tp1 match {
+      case tp1: TypeRef => tp1.symbol.is(Package) || isStaticPathOK(tp1.prefix)
+      case tp1: TermRef => tp1.symbol.isStaticOwner || isStaticPathOK(tp1.prefix)
+      case tp1: ThisType => tp1.tref.symbol.is(Package) || tp1.tref.symbol.is(Module)
+      case tp1: AppliedType => isStaticPathOK(tp1.tycon)
+      case _: SkolemType => true
+      case _ => false
+  }
+
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: Some indentation seems to be off here.

File: compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala]: I used this version it but made a couple of extra changes to make all the tests pass

**Pull request merged**